engine/src/main/java/org/pentaho/di/trans/step/BaseStep.java

```diff
@@ -1750,33 +1751,33 @@ private void verifyRejectionRates() {
    */
   @VisibleForTesting
   RowSet currentInputStream() {
-    return inputRowSets.get( currentInputRowSetNr );
+    synchronized ( inputRowSetsLock ) {
+      return inputRowSets.get( currentInputRowSetNr );
+    }
   }
 
   /**
    * Find the next not-finished input-stream... in_handling says which one...
    */
   private void nextInputStream() {
-    synchronized ( inputRowSets ) {
-      blockPointer = 0;
+    blockPointer = 0;
 
-      int streams = inputRowSets.size();
+    int streams = inputRowSets.size();
```

Comment:
Same comment as above - inputRowSets can be mutated above, so likely need a readWriteLock.lock here.

Reply:
Same here - private method only used in synchronized context.